frame/system/src/lib.rs

```diff
@@ -1068,6 +1069,13 @@ impl<T: Config> Pallet<T> {
 		Account::<T>::contains_key(who)
 	}
 
+	fn do_set_code(code: Vec<u8>) -> DispatchResult {
+		T::OnSetCode::set_code(code)?;
+		Self::deposit_log(generic::DigestItem::RuntimeUpdated);
```

Comment:
This will not work for Parachains.

Parachains have a special `set_code` implementation that will not update `:code` here directly. It will first queue the update and update it when it is allowed by the relay chain.

Reply:
Where can we find the queue implementation? Any chance the actual storage write can go through some function that would emit the log?

Reply:
Yeah we can do this.

https://github.com/paritytech/cumulus/blob/master/pallets/parachain-system/src/lib.rs#L943

Reply:
I've exposed a helper function to update the storage and emit event + log. The idea is that this function will be used by parachain code implementation and is used by the default implementation of `OnSetCode` for unit type.